### PR TITLE
Dynamic descriptions of card effects

### PR DESCRIPTION
**Pull request by ariel215**
Dynamic descriptions of card effects. 

The intent is for these to be used when a card is on top of the queue, in order to show what it will _actually_ do.

### CODE REVIEW

File: Assets/Scripts/Effects/Description.cs [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+// This is a liiiittle bit of a cop-out to avoid 
+// caring about the context an effect appears in
+
+using System.Collections.Generic;
+using Effects;
+using Cards;
+
+// Describes damage, healing, armor, Deciever cost --
+// everything that happens in Effects.Effect
+public class EffectDescription
```
Comment (infraredsqy) [inline Assets/Scripts/Effects/Description.cs:15]: Since you implemented the parser in this way (not what I expected). Is there any scene showing the update strings to check if it is working? According to Rider, none of the functions is actually called

File: Assets/Scripts/Effects/Description.cs [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+        EnemyStats = new Stats(0, 0);
+        SelfStats = new Stats(0, 0);
+        NotAmplified = false;
+        DiscardDecievers = false;
+    }
+
+    // Figure out how much damage / healing / armor an effect is going to affect
+    private static void GetStats(Stats statBlock, List<Card> discardPile, int cardsInHand, Effect effect)
+    {
+        switch (effect.type)
```
Comment (infraredsqy) [inline Assets/Scripts/Effects/Description.cs:86]: Say I added 4 more effects, I need to update this 'switch' block as well as the actual 'ability' block... not a great plan

**Review by infraredsqy — DISMISSED** [review]
You basically wrote another damage calculator in order to parse, but I do want to mention that having two pieces of code doing a similar job is redundant, and hard for maintenance & function expansion.
The reason I asked for a parser is that: 
 - The player needs it according to playtesting
 - Internal function doing similar work (e.g. System.Linq in .NET) doesn't really work here as a dynamic parser

What I want is:
 - Figure out pairs of "{" and  "}" similar to an equation parser (no bracket-inside-bracket, so should be relatively easy)
 - Match those pairs to the effects/buffs in effectlist/bufflist using a syntac system you defined (e.g. {effects[1].amount} goes to the 2nd amount of effect of a card)
 - And at last, multiply the amount with the streakCount, so the only calculation needed here is just a multiplication.

PM me if you have any questions.

**Comment by ariel215** [comment]
To address your concerns:

1. I've refactored the codebase to minimize duplication, though some additional work could be done in this regard.
2. I think there's some merit to having the card say what it does in the abstract when it's in your hand, and once it's played say only what it's going to do if you end the turn here. In particular, the minimum streak information is only relevant while in your hand; once you play the card, either its streak has procc'ed or it hasn't
3. Displaying text based on what the cards are programmed to do is better for purpose of debugging.
4. On this branch the descriptions are fully integrated into the card system, and are used to set the text of the top card in the stack i.e. the card that is going to be played. 

In conclusion, due perhaps to some amount of inadequate communication this is the system I've written.  I think it does a good job.

File: Assets/Scripts/Effects/Description.cs (hunk 2 above)
Comment (ariel215) [reply Assets/Scripts/Effects/Description.cs:86]: (A) If they're at all complicated you'd need to change this file anyway, (B) there's one week left what are we doing adding four more effects

Comment (infraredsqy) [reply Assets/Scripts/Effects/Description.cs:86]: Programmers want to do 'general' things. Write the code in a pure 'String Parser' way would make adding and maintaining existing code much easier.

... but anyway. still checking

File: Assets/Scripts/Effects/Description.cs [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+            {
+                desc += "Deal " + EnemyDamageStr + ".";
+            }
+        }
+        if (SelfStatsHighBound.Damage < 0)
+        {
+            desc += "Gain " + SelfHealStr;
+            if (EnemyStatsHighBound.Damage < 0)
+            {
+                desc += "and enemy gains" + EnemyHealStr + ".";
```
Comment (infraredsqy) [inline Assets/Scripts/Effects/Description.cs]: Some of the strings are missing spaces that making the description weird

Comment (ariel215) [reply Assets/Scripts/Effects/Description.cs]: Fixed now

**Comment by infraredsqy** [comment]
Since you are working on this:
* The final implementation of the dynamic description probably goes through a UI 'hint speech bubble' instead of changing the card directly.
* The strings will be shown on a floating speech bubble, which means that...
  * You can either implement the bubble itself (this asks for a bunch of coding details), or you can leave me with a string API so that I can call the function whenever needed.

**Comment by ariel215** [comment]
@QifanYang-bw I've added public methods EffectPreview(), BuffPreview(), and Preview() to CardRender which return strings  describing the net effects, buffs, and both together respectively.

I've left in the implementation where it replaces the text of the top card in the queue for now; to get rid of it, remove CardRender.LateUpdate().
